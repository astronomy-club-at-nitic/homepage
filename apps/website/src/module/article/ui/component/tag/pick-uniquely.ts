export const pickUniquely = <TCandidate>(seed: string, candidates: Array<TCandidate>) => {
  // If there are no candidates, return undefined
  if (candidates.length === 0) {
    return undefined;
  }
  // Calculate the sum of the Unicode code points of the seed string
  const seedSum = seed.split('').reduce((sum, char) => sum + char.charCodeAt(0), 0);

  // Pick a candidate based on the seed sum
  // The index must be zero or a positive integer
  const index = Math.abs(Math.round(seedSum)) % candidates.length;

  return candidates[index];
};
