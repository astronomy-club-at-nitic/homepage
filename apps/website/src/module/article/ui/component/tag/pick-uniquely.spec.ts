import { pickUniquely } from './pick-uniquely';

describe('pickUniquely', () => {
  it('should return undefined if there are no candidates', () => {
    const result = pickUniquely('seed', []);
    expect(result).toBeUndefined();
  });

  it('should return a candidate', () => {
    const result = pickUniquely('seed', ['candidate']);
    expect(result).toBe('candidate');
  });

  it('should return a candidate matching the snapshot', () => {
    const result = pickUniquely('seed', ['c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7', 'c8']);
    expect(result).not.toBeUndefined();
    expect(result).toMatchSnapshot();
  });
});
