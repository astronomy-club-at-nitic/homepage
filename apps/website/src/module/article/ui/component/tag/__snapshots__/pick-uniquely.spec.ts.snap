// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pickUniquely should return a candidate matching the snapshot 1`] = `"c2"`;
